aliasing fixes

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #66977

Fix for https://github.com/pytorch/pytorch/issues/47218

More context is in original PR here: https://github.com/pytorch/pytorch/pull/20556

Differential Revision: [D31935573](https://our.internmc.facebook.com/intern/diff/D31935573)

Commit:
aliasing fixes

[ghstack-poisoned]

Commit:
Update on "aliasing fixes"




[ghstack-poisoned]

dzhulgakov reviewed (commented):
Update native/README.md with what this syntax means?

albanD reviewed (commented):
Why do you need a new syntax here?
The previous syntax has the same meaning no?

Commit:
Update on "aliasing fixes"


Fix for https://github.com/pytorch/pytorch/issues/47218

More context is in original PR here: https://github.com/pytorch/pytorch/pull/20556


[ghstack-poisoned]

Commit:
Update on "aliasing fixes"


Fix for https://github.com/pytorch/pytorch/issues/47218

More context is in original PR here: https://github.com/pytorch/pytorch/pull/20556


[ghstack-poisoned]

Commit:
Update on "aliasing fixes"


Fix for https://github.com/pytorch/pytorch/issues/47218

More context is in original PR here: https://github.com/pytorch/pytorch/pull/20556

Differential Revision: [D31935573](https://our.internmc.facebook.com/intern/diff/D31935573)

[ghstack-poisoned]

eellison reviewed (commented):
I updated and extended the views logic, but we should check in about refactoring the logic. some of the assumptions, like the fact that an alias annotation is present means there's a view, doesn't hold up with the aliasing grammar. 

For instance: `Tensor(a) -> Tensor(b)` should be fine but trigger a view here. I think what we need to do is factor out the reader of schemas in alias_analysis to be seprately queryable.

soulitzer reviewed (commented):
Thanks for updating the autograd fallback logic to work with the new aliasing grammar. The changes look good, I just had some small comments.

eellison commented on torch/csrc/autograd/autograd_not_implemented_fallback.cpp:
```diff
@@ -41,6 +41,10 @@ void _foreach_tensor(
   }
 }
 
+bool isTensorOfLists(const at::TypePtr& type) {
```
There is a `isTensorList` method on ivalues of the corresponding other method you are talking about

Commit:
Update on "aliasing fixes"


Fix for https://github.com/pytorch/pytorch/issues/47218

More context is in original PR here: https://github.com/pytorch/pytorch/pull/20556

Differential Revision: [D31935573](https://our.internmc.facebook.com/intern/diff/D31935573)

[ghstack-poisoned]

albanD reviewed (commented):
I didn't check the not implemented fallback in details as @soulitzer knows that better.
Is there any perf implication of these modifications though?

eellison commented on aten/src/ATen/native/README.md:
```diff
@@ -250,6 +250,13 @@ There is also another situation in which we use annotations, namely views.
+When a Tensor views are contained in a Tensor list, we need to represent that the output list
+contains Tensors that alias the input.
+  - `func: chunk(Tensor(a -> *) self, int chunks, int dim=0) -> Tensor[]`
+We assume lists contain memory which aliases the heap, so in order to correctly set up the aliasing
```
I don't want to re-duplicate the alias docs everywhere... will add link/ref

eellison commented on tools/codegen/model.py:
```diff
@@ -883,19 +885,30 @@ class Annotation:
+        if becomes_wildcard_index != -1:
+            after_set = "*"
+            # TODO: im not good enough with regexes to ignore -> *
+            m = re.match(r'^([a-z])(!?)(!?)$', ann[:becomes_wildcard_index] + ann[becomes_wildcard_index + len(" -> *"):])
```
foreach functions dont have this problem because the input lists are assumed to be in the heap

eellison replied:
Yea, in this PR
> We assume lists contain memory which aliases the heap

I think there are a number of follow up issues to unify and/improve the aliasing specification and particularly the conformance of both autograd's understanding of it and JIT's. I knew that there was a re-implementation of aliasing parsing happening in python but was not really aware of the details...

eellison reviewed (commented):
@albanD haven't tested perf yet but the long term plan would be to cache per operator (see https://github.com/issues/). Inevitably as we add more of the aliasing schema parsing to this pathway (see https://github.com/pytorch/pytorch/issues/30526) perf will slightly degrade, but IMO for that we should add a cache not over-optimize the parsing.

Commit:
Update on "aliasing fixes"


Fix for https://github.com/pytorch/pytorch/issues/47218

More context is in original PR here: https://github.com/pytorch/pytorch/pull/20556

Differential Revision: [D31935573](https://our.internmc.facebook.com/intern/diff/D31935573)

[ghstack-poisoned]

Commit:
Update on "aliasing fixes"


Fix for https://github.com/pytorch/pytorch/issues/47218

More context is in original PR here: https://github.com/pytorch/pytorch/pull/20556

Differential Revision: [D31935573](https://our.internmc.facebook.com/intern/diff/D31935573)

[ghstack-poisoned]

bdhirsh commented:
There's an `is_view_op` method defined in `model.py` here: https://github.com/pytorch/pytorch/pull/66977/files#diff-a34ee043aa7bde7b608aa03bdd3ff6d24af2fea53f96c01be304034c30a8bfcdR487 that I think you'll need to change- right now it determines aliasing ops based on whether the output tensor has alias annotations.

By the way, what's the reason that we don't need/want alias annotations on the return argument anymore for those multi-tensor-output views? (e.g. split)

eellison commented:
@bdhirsh the internal aliasing syntax that was used wasn't even being parsed, (e.g. Tensor(a)[] == Tensor()[]). it was just the fact that the output has an alias annotation at all that affected the logic. ideally, having an alias annotation of a different shouldn't affect the parsing, that's not part of the grammar specification. e.g. Tensor(a) -> Tensor(b) should be the same as Tensor -> Tensor.

bdhirsh commented:
>  e.g. Tensor(a) -> Tensor(b) should be the same as Tensor -> Tensor.

Hmm maybe I'm not following. Doesn't that imply that:

`Tensor(a) -> Tensor` should be the same as `Tensor -> Tensor` too?

Then for this schema:
`(Tensor(a -> *) -> Tensor[]`

Is the `a -> *` annotation for the argument meant to imply that the return aliases the argument? That's not super intuitive to me, but I also don't know as much about how torchscript deals with aliasing.

Commit:
Update on "aliasing fixes"


Fix for https://github.com/pytorch/pytorch/issues/47218

More context is in original PR here: https://github.com/pytorch/pytorch/pull/20556

Differential Revision: [D31935573](https://our.internmc.facebook.com/intern/diff/D31935573)

[ghstack-poisoned]

Krovatkin reviewed (approved):
LGTM

eellison commented:
> Tensor(a) -> Tensor should be the same as Tensor -> Tensor too?

Yes it does. Per aliasing specification.  

> (Tensor(a -> *) -> Tensor[]

I agree that it's not super intuitive, but see comment above:

...

Yea, I am not saying this is the ideal syntax but updating it would require:
- parsing and storing contained alias sets, which we would have to change in two different parsers because the original aliasing parser is now duplicated in python (thus the changes i am making here)
and then either
- tracking the contained elements of containers and not automatically be the wildcard set. that would be a huge change and involve a number of schema, aliasdb, and memory dag changes
- logic to see if an annotated alias set has now been used in a contained alias set which would imply that it has entered the wildcard set.
I think the 2 would be a good medium term change but together is a pretty good changeset for what is a bug and effects memory management etc in SR, autodiff correctness for JIT, etc, and this PR fixes the bug in a way that makes the schemas compatible with the other existing schemas today.

Commit:
Update on "aliasing fixes"


Fix for https://github.com/pytorch/pytorch/issues/47218

More context is in original PR here: https://github.com/pytorch/pytorch/pull/20556

Differential Revision: [D31935573](https://our.internmc.facebook.com/intern/diff/D31935573)

[ghstack-poisoned]

Commit:
Update on "aliasing fixes"


Fix for https://github.com/pytorch/pytorch/issues/47218

More context is in original PR here: https://github.com/pytorch/pytorch/pull/20556

Differential Revision: [D31935573](https://our.internmc.facebook.com/intern/diff/D31935573)

[ghstack-poisoned]

Commit:
Update on "aliasing fixes"


Fix for https://github.com/pytorch/pytorch/issues/47218

More context is in original PR here: https://github.com/pytorch/pytorch/pull/20556

Differential Revision: [D31935573](https://our.internmc.facebook.com/intern/diff/D31935573)

[ghstack-poisoned]

Commit:
Update on "aliasing fixes"


Fix for https://github.com/pytorch/pytorch/issues/47218

More context is in original PR here: https://github.com/pytorch/pytorch/pull/20556

Differential Revision: [D31935573](https://our.internmc.facebook.com/intern/diff/D31935573)

[ghstack-poisoned]

bdhirsh commented:
@eellison let me know if you disagree, but to me `Tensor(a) -> Tensor(a)[]` seems like the "correct" frontend alias annotations for an op like `torch.split`. The downside of changing the annotations in the frontend in this way is that it's a breaking change for any code that relies on the existing aliasing semantics, including external code:
- the codegen (manually fixed here)
- the autograd not-implemented fallback (manually fixed here)
- any external hardware vendors that rely on the existing aliasing semantics externally (I think for example, MLC might rely on this cc @albanD)

These all would no longer be able to trust the alias annotations of the return arguments, which doesn't feel like the correct end-state.

If the existing alias annotations result in a bug for torchscript/SR, then that sound like something we should try to work around directly in the torchscript alias handling?

> yea, I am not saying this is the ideal syntax but updating it would require...

You pointed out that "really fixing it" would require a pretty huge change. If we agree that the existing bug is in torchscript's aliasing machinery, then I wonder if we can add a simpler bandaid by maybe transforming the aliasing annotations under the hood for torchscript? E.g. literally converting `Tensor(a) -> Tensor(a)[]` to `Tensor(a -> *) -> Tensor[]` at the time that torchscript parses aliasing annotations.

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31935573).

eellison commented:
> Tensor(a) -> Tensor(a)[] seems like the "correct" frontend

Yea, i agree with that, but this annotation is being ignored both in every parser that exists right now. Something like `Tensor(a) -> Tensor(b)[]` would result in the same exact behavior with respect to the current handling. The `Tensor(a)[]` is being completely ignored, and instead the only thing that is being read is the presence of an annotation at all. 

If we want to improve the aliasing grammar's ability to specify things, I'm all for it, but this annotation syntax was added without the corresponding changes to the grammar that would actually enable the ideal state that you are talking about.

albanD commented:
> Yea, i agree with that, but this annotation is being ignored both in every parser that exists right now (at least within core). Something like Tensor(a) -> Tensor(b)[] would result in the same exact behavior with respect to the current handling. The Tensor(a)[] is being completely ignored, and instead the only thing that is being read is the presence of an annotation at all.

Do you mean that all the existing fallback are buggy and not doing the right thing in the context of Tensor[] ?
A quick test of the conjugate view one seems to show that it works just fine right now:
```python
import torch

a = torch.rand(3, 3, dtype=torch.complex64)
print(a)

b = a.conj()

c = b.unbind()
print(c)
```

eellison commented:
@albanD the aliasing specification within native_functions.yaml is “if there is an annotation, assume things alias”. Which seems to work ! But is not the same as implementing parsing of contained alias sets as described by Brian above. 

Also, see operators like https://github.com/pytorch/pytorch/issues/30526

bdhirsh commented:
From talking offline with @albanD, I think we agree that the main constraint is that we don't want to BC-break any boxed fallback code, including code from external backends. I think this PR as is actually breaks the pytorch/xla CPU fallback (although it doesn't show up in CI) [here](https://github.com/pytorch/pytorch/blob/4d601a1c362a300450ecc3204eeca24cb17c0b38/aten/src/ATen/native/CPUFallback.cpp#L138). We also don't have a good way of checking / knowing what external code might broken. So I think the main contraint is that return arguments should continue show up as aliasing if they used to show up as aliasing.

> I can special case the aliasing behavior within alias_analysis.cpp if we want, fine, but that not a great state to be in IMO as it makes custom operators now divergent from native_functions.yaml ones.

Where does the divergent behavior come from for custom ops? My first thought would be that everything boils down to a `TORCH_LIBRARY` call (both custom and native ops), which goes through the same schema parsing machinery.

eellison commented:
In this parser we do check that the alias_info are matching in this branch https://github.com/pytorch/pytorch/blob/4d601a1c362a300450ecc3204eeca24cb17c0b38/aten/src/ATen/native/CPUFallback.cpp#L151

but then do not check the same thing in the following branch https://github.com/pytorch/pytorch/blob/4d601a1c362a300450ecc3204eeca24cb17c0b38/aten/src/ATen/native/CPUFallback.cpp#L163 

I think that's because this is encoding the special meaning which is that outputs which have annotations are always assumed to alias, without parsing the contained `(a)`. I can retain that by adding the `Tensor(a)[]` output tensor annotation.

> Where does the divergent behavior come from for custom ops? My first thought would be that everything boils down to a TORCH_LIBRARY

If i were to add a custom aliasing logic in alias_analysis.cpp instead of making the change to the schemas as reflected here, that's where the divergence would come from.

Commit:
Update on "aliasing fixes"


Fix for https://github.com/pytorch/pytorch/issues/47218

More context is in original PR here: https://github.com/pytorch/pytorch/pull/20556

Differential Revision: [D31935573](https://our.internmc.facebook.com/intern/diff/D31935573)

[ghstack-poisoned]

Commit:
Update on "aliasing fixes"


Fix for https://github.com/pytorch/pytorch/issues/47218

More context is in original PR here: https://github.com/pytorch/pytorch/pull/20556

Differential Revision: [D31935573](https://our.internmc.facebook.com/intern/diff/D31935573)

[ghstack-poisoned]

Commit:
Update on "aliasing fixes"


Fix for https://github.com/pytorch/pytorch/issues/47218

More context is in original PR here: https://github.com/pytorch/pytorch/pull/20556

Differential Revision: [D31935573](https://our.internmc.facebook.com/intern/diff/D31935573)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31935573).

Commit:
Update on "aliasing fixes"


Fix for https://github.com/pytorch/pytorch/issues/47218

More context is in original PR here: https://github.com/pytorch/pytorch/pull/20556

Differential Revision: [D31935573](https://our.internmc.facebook.com/intern/diff/D31935573)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31935573).

albanD commented on aten/src/ATen/native/README.md:
```diff
@@ -250,6 +250,13 @@ There is also another situation in which we use annotations, namely views.
 
+When a Tensor views are contained in a Tensor list, we need to represent that the output list
+contains Tensors that alias the input.
+  - `func: chunk(Tensor(a -> *) self, int chunks, int dim=0) -> Tensor[]`
```
nit the output schema needs updating?

albanD commented on torch/csrc/autograd/autograd_not_implemented_fallback.cpp:
```diff
@@ -64,36 +83,72 @@ void autogradNotImplementedFallbackImpl(const c10::OperatorHandle& op, c10::Disp
+  int aliased_input_idx = -1;
+  int aliased_output_idx = -1;
+
+  // if a Tensor is annotated with (a -> *) that implies that all lists of
```
Can the changes to this file be reverted now?

Commit:
Update on "aliasing fixes"


Fix for https://github.com/pytorch/pytorch/issues/47218

More context is in original PR here: https://github.com/pytorch/pytorch/pull/20556

Differential Revision: [D31935573](https://our.internmc.facebook.com/intern/diff/D31935573)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31935573).

Commit:
Update on "aliasing fixes"


Fix for https://github.com/pytorch/pytorch/issues/47218

More context is in original PR here: https://github.com/pytorch/pytorch/pull/20556

Differential Revision: [D31935573](https://our.internmc.facebook.com/intern/diff/D31935573)

[ghstack-poisoned]

Commit:
Update on "aliasing fixes"


Fix for https://github.com/pytorch/pytorch/issues/47218

More context is in original PR here: https://github.com/pytorch/pytorch/pull/20556

Differential Revision: [D31935573](https://our.internmc.facebook.com/intern/diff/D31935573)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31935573).

Commit:
Update on "aliasing fixes"


Fix for https://github.com/pytorch/pytorch/issues/47218

More context is in original PR here: https://github.com/pytorch/pytorch/pull/20556

Differential Revision: [D31935573](https://our.internmc.facebook.com/intern/diff/D31935573)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31935573).

albanD commented on test/cpp/api/autograd.cpp:
```diff
@@ -1220,4 +1220,4 @@ TEST(TestAutogradNotImplementedFallback, TensorlistOp) {
 // test_return_leaf
 // test_anomaly_detect_nan
-// test_no_grad_copy
+// test_no_grad_copy
```
nit missing newline?

albanD reviewed (approved):
LGTM
Some newlines at the end of files to make lint happy only.

Commit:
Update on "aliasing fixes"


Fix for https://github.com/pytorch/pytorch/issues/47218

More context is in original PR here: https://github.com/pytorch/pytorch/pull/20556

Differential Revision: [D31935573](https://our.internmc.facebook.com/intern/diff/D31935573)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31935573).

Commit:
Update on "aliasing fixes"


Fix for https://github.com/pytorch/pytorch/issues/47218

More context is in original PR here: https://github.com/pytorch/pytorch/pull/20556

Differential Revision: [D31935573](https://our.internmc.facebook.com/intern/diff/D31935573)

[ghstack-poisoned]

Commit:
Update on "aliasing fixes"


Fix for https://github.com/pytorch/pytorch/issues/47218

More context is in original PR here: https://github.com/pytorch/pytorch/pull/20556

Differential Revision: [D31935573](https://our.internmc.facebook.com/intern/diff/D31935573)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31935573).
